File: 2018/01/solution.py
from functools import reduce
from itertools import cycle

def parse_input():
    with open('input.txt') as f:
        return [*map(int, f.readlines())]

def part1(frequencies):
    return reduce(
        lambda x, y: x + y,
        frequencies,
        0
    )

def part2(frequencies):
    cumulative = 0
    seen = set([0])
    for frequency in cycle(frequencies):
        cumulative += frequency
        if cumulative in seen:
            return cumulative
        seen.add(cumulative)

frequencies = parse_input()
print(part1(frequencies))
print(part2(frequencies))

File: 2018/01/solution.rs
use std::collections::HashSet;
use std::fs;

fn main() {
    let input = parse_input();
    println!("{}", part1(&input));
    println!("{}", part2(&input));
}

fn parse_input() -> Vec<i32> {
    fs::read_to_string("input.txt")
        .unwrap()
        .lines()
        .map(|line| line.parse().unwrap())
        .collect()
}

fn part1(input: &Vec<i32>) -> i32 {
    input.iter().fold(0, |acc, frequency| acc + frequency)
}

fn part2(input: &Vec<i32>) -> i32 {
    let mut cumulative = 0;
    let mut seen = HashSet::new();
    for frequency in input.iter().cycle() {
        cumulative += frequency;
        if seen.contains(&cumulative) {
            return cumulative;
        }
        seen.insert(cumulative);
    }
    panic!()
}
